### PR TITLE
Fix cloud config updates & docker issue

### LINKED ISSUE #560: add --rm flag to cloud-init image preload
We have spotted thousands of finished containers in the shoot nodes as a result of the cloud-init unit image pre-load. We are talking about 30000 containers in a week, which is pretty high and should not be happening, impacting also in kubernetes not being able to purge images, and node disk space usage

So far, I think it can be easily solved adding the flag `--rm` to the `docker run` command on this line:
https://github.com/gardener/gardener/blob/e146c72eea2d9fd08b3885dfff6a9cd54199bc06/charts/shoot-cloud-config/charts/downloader/templates/cloud-config/_download-cloud-config.sh#L39

### PR DESCRIPTION
**Pull request by adracus**
**What this PR does / why we need it**:
Fix orphan containers being created during regular cloud config sync.
Enable rollout of cloud config fixes via secret content update.

**Which issue(s) this PR fixes**:
Fixes #560

**Special notes for your reviewer**:

**Release note**:

```noteworthy user
⚠️ An issue with orphaned containers being created during the node's cloud config syncs has been fixed. Deploying this change requires a rolling update of all worker nodes of all shoot clusters.
```

### CODE REVIEW

**Comment by vlerenc** [comment]
Very good, another critical issue resolved. I am wondering how we could protect us from this in the future. I hope we can write integration tests and make sanity checks afterwards like scanning the logs for warnings and errors or check certain things like how many outbound calls (to the infrastructure where made), were there any left overs, e.g. IaaS resources like volumes, and things like to the pods/containers match the actual containers on the file system (there are a ton of more things to check of course). I know that fixing the issue was of highest priority, but maybe we can hve a chat after the patch release, how it can become easy/easier to write/add such checks in the future. The easier it is, the more often we will do it, which is in the end the only thing that can protect us, would you agree?

File: charts/shoot-cloud-config/charts/downloader/templates/cloud-config/_download-cloud-config.sh [hunk 1]
```diff
@@ -24,79 +22,10 @@ function kubectl() {
     kubectl "$@"
```
Comment (rfranzke) [inline charts/shoot-cloud-config/charts/downloader/templates/cloud-config/_download-cloud-config.sh:14]: Can we remove the volumes mounts

```bash
     -v "$DIR_CLOUDCONFIG"/:"$DIR_CLOUDCONFIG" \
     -v "$DIR_CLOUDCONFIG_DOWNLOADER"/:"$DIR_CLOUDCONFIG_DOWNLOADER" \
```

in the `kubectl` function now?

Comment (rfranzke) [inline charts/shoot-cloud-config/charts/downloader/templates/cloud-config/_download-cloud-config.sh:14]: Let's bump hyperkube to `v1.12.3` if we now need to touch the script anways.

File: charts/shoot-cloud-config/charts/downloader/templates/cloud-config/_download-cloud-config.sh [hunk 2]
```diff
@@ -8,8 +8,6 @@ DIR_CLOUDCONFIG="/var/run/coreos"
 DIR_KUBELET="/var/lib/kubelet"
 
 PATH_KUBECONFIG="$DIR_CLOUDCONFIG_DOWNLOADER/kubeconfig"
-PATH_CLOUDCONFIG="$DIR_CLOUDCONFIG/cloud_config.yml"
-PATH_CLOUDCONFIG_OLD="$DIR_CLOUDCONFIG/cloud_config.old.yml"
 
 mkdir -p "$DIR_CLOUDCONFIG" "$DIR_KUBELET"
```
Comment (rfranzke) [inline charts/shoot-cloud-config/charts/downloader/templates/cloud-config/_download-cloud-config.sh]: The downloaded script can do it on its own, right?
Can we remove this (same applies for the used variables)?

File: charts/shoot-cloud-config/charts/original/templates/_cloud-config-script.tpl [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+  /usr/bin/docker run --rm "$@"
+}
+
+function kubectl() {
+  docker-run \
+    --net host \
+    -v "$DIR_CLOUDCONFIG"/:"$DIR_CLOUDCONFIG" \
+    -v "$DIR_CLOUDCONFIG_DOWNLOADER"/:"$DIR_CLOUDCONFIG_DOWNLOADER" \
+    -e "KUBECONFIG=$PATH_KUBECONFIG" \
+    k8s.gcr.io/hyperkube:v1.9.3 \
```
Comment (rfranzke) [inline charts/shoot-cloud-config/charts/original/templates/_cloud-config-script.tpl]: Let's bump hyperkube to `v1.12.3` if we now need to touch the script anways.

File: pkg/operation/hybridbotanist/cloud_config.go [hunk 4]
```diff
@@ -78,10 +78,11 @@ func (b *HybridBotanist) generateCloudConfigChart() (*chartrenderer.RenderedChar
 			},
 			"version": b.Shoot.Info.Spec.Kubernetes.Version,
 		},
-		"workers": workers,
+		"workers":    workers,
+		"kubeconfig": b.Secrets["cloud-config-downloader"].Data["kubeconfig"],
```
Comment (rfranzke) [inline pkg/operation/hybridbotanist/cloud_config.go]: Why is this here? Seems not to be used anywhere, nor is it part of the `values.yaml`.

**Review by rfranzke — CHANGES_REQUESTED** [review]

File: pkg/operation/hybridbotanist/cloud_config.go [hunk 5]
```diff
@@ -78,10 +78,11 @@ func (b *HybridBotanist) generateCloudConfigChart() (*chartrenderer.RenderedChar
...
 	}
 
-	config, err = b.InjectImages(config, b.ShootVersion(), b.ShootVersion(), common.HyperkubeImageName)
+	config, err = b.InjectImages(config, b.ShootVersion(), b.ShootVersion(), common.RubyImageName, common.HyperkubeImageName)
```
Comment (timuthy) [inline pkg/operation/hybridbotanist/cloud_config.go:84]: Just for sake of integrity, can you add the RubyImage to https://github.com/gardener/gardener/blob/90c0d6aa6150f3a265dcb5b77b2426a2f9947817/charts/shoot-cloud-config/charts/original/values.yaml?

**Review by timuthy — CHANGES_REQUESTED** [review]

**Review by timuthy — APPROVED** [review]

**Comment by adracus** [comment]
@vlerenc agree regarding your idea, in general we need more well-typed and well-tested code where we know about the side effects. However, in particular for bash scripts I don't immediately see a handful of measures other than replacing them by proper binaries (which lead to other required builds / other problems?). As said, let's have a chat afterwards.

**Review by rfranzke — APPROVED** [review]
/lgtm

**Comment by vlerenc** [comment]
Yes, unit tests for bash scripts suck and writing both in Go makes things certainly better/easier. But I am thinking about something different here: I am concerned about side effects and finding a way to properly test for those. This is unrelated to the flow, whether there were bash scripts or binaries at work. It's about finding a way to plug in tests that take a look at the system, while it is running or towards the end, and raise test failures. E.g. we may decide not to want warnings and errors in our Gardener API/Ctr logs after running the test or we may decide to inspect the pods and their state (all running, no restarts) or we may decide to inspect the Docker daemon before we destroy the machine and make sure there are no surprising Docker containers (especially, if we ran into that problem once, so that we never again run into the problem). Only when we reach this level of testing, beyond unit tests, we will get better/reach higher quality. But that's hard, just saying, that we need that eventually, especially because of the tremendous pull we see.

**Pull request merged**